test/xbgpu/test_engine.py

```diff
@@ -821,77 +818,91 @@ async def _send_data(
+                        == accumulation_index * timestamp_step * corrprod_output.heap_accumulation_threshold
+                    ), (
+                        "Output timestamp is not correct. "
+                        f"Expected: "
+                        f"{hex(accumulation_index * timestamp_step * corrprod_output.heap_accumulation_threshold)}, "
+                        f"actual: {hex(ig_recv['timestamp'].value)}."
+                    )
 
-                assert ig_recv["frequency"].value == frequency, (
-                    "Output channel offset not correct. "
-                    f"Expected: {frequency}, "
-                    f"actual: {ig_recv['frequency'].value}."
-                )
+                    assert ig_recv["frequency"].value == frequency, (
+                        "Output channel offset not correct. "
+                        f"Expected: {frequency}, "
+                        f"actual: {ig_recv['frequency'].value}."
+                    )
 
-                corrprod_results[i][j] = ig_recv["xeng_raw"].value
+                    corrprod_results[corrprod_output.name][j] = ig_recv["xeng_raw"].value
+            except spead2.Stopped:
+                pass
 
-        beam_results = np.zeros(
-            (len(beam_outputs), len(batch_indices), n_channels_per_substream, n_spectra_per_heap, COMPLEX),
-            bsend.SEND_DTYPE,
-        )
-        for i in range(len(beam_outputs)):
-            stream = spead2.recv.asyncio.Stream(out_tp, out_config)
-            stream.add_inproc_reader(mock_send_stream[i + len(corrprod_outputs)])
-            # It is expected that the first packet will be a descriptor.
-            ig_recv = spead2.ItemGroup()
-            heap = await stream.get()
-            items = ig_recv.update(heap)
-            assert len(list(items.values())) == 0, "This heap contains item values not just the expected descriptors."
-
-            for j, index in enumerate(batch_indices):
+        beam_results = {
+            beam_output.name: np.zeros(
+                (len(batch_indices), n_channels_per_substream, n_spectra_per_heap, COMPLEX),
+                bsend.SEND_DTYPE,
+            )
+            for beam_output in beam_outputs
+        }
+
+        for i, beam_output in enumerate(beam_outputs):
+            try:
+                stream = spead2.recv.asyncio.Stream(out_tp, out_config)
+                stream.add_inproc_reader(mock_send_stream[i + len(corrprod_outputs)])
+                # It is expected that the first packet will be a descriptor.
+                ig_recv = spead2.ItemGroup()
                 heap = await stream.get()
-                while (updated_items := set(ig_recv.update(heap))) == set():
-                    # Test has gone on long enough that we've received another descriptor
-                    heap = await stream.get()
+                items = ig_recv.update(heap)
+                assert (
+                    len(list(items.values())) == 0
```

Comment:
I realise you haven't actually changed this code, just reflowed it, but I think this could just be
```suggestion
                    len(items) == 0
```
That would probably allow it to go back to being a single line instead of awkwardly wrapped.